File: graphql-go/internal/query/sql_builder.go
package query

import (
	"fmt"
	"strings"
)

const defaultInitialBufferCapacity = 1024

func NewSqlVisitor() SqlVisitor {
	return SqlVisitor{}
}

func NewSqlVisitorWithJoinedTable(jt *JoinedTable) SqlVisitor {
	return SqlVisitor{
		joinedTable: jt,
	}
}

type SqlVisitor struct {
	joinedTable  *JoinedTable
	nextTable    int
	currentTable []int
	query        strings.Builder
	params       []any
}

func (s *SqlVisitor) Translate(filter TableFilter, sel []string) (string, []any) {
	s.query.Grow(defaultInitialBufferCapacity)
	currentTableId := s.pushNewTable()

	s.query.WriteString("SELECT ")
	if len(sel) == 0 {
		s.query.WriteString("*")
	} else {
		for idx, sl := range sel {
			if idx > 0 {
				s.query.WriteString(", ")
			}

			s.query.WriteString(fmt.Sprintf("T%d.%s", currentTableId, sl))
		}
	}

	s.query.WriteString(" FROM ")
	if s.joinedTable != nil {
		s.query.WriteString(fmt.Sprintf(
			"%s AS J INNER JOIN %s AS T%d ON J.%s = T%d.%s",
			s.joinedTable.JoinTable,
			filter.Table,
			currentTableId,
			s.joinedTable.JoinTableJoinCol,
			currentTableId,
			s.joinedTable.DataTableJoinCol,
		))
	} else {
		s.query.WriteString(fmt.Sprintf("%s AS T%d", filter.Table, currentTableId))
	}

	if !filter.FilterGroup.IsEmpty() || s.joinedTable != nil {
		s.query.WriteString(" WHERE ")
	}

	if !filter.FilterGroup.IsEmpty() {
		filter.Accept(s)
	}

	if s.joinedTable != nil {
		if !filter.FilterGroup.IsEmpty() {
			s.query.WriteString(" AND ")
		}

		param := s.addParam(s.joinedTable.JoinTableFilterColVal)
		s.query.WriteString(fmt.Sprintf("J.%s = $%d", s.joinedTable.JoinTableFilterCol, param))
	}

	return s.query.String(), s.params
}

func (s *SqlVisitor) pushNewTable() int {
	s.currentTable = append(s.currentTable, s.nextTable)
	s.nextTable += 1
	return s.nextTable - 1
}

func (s *SqlVisitor) popOldTable() {
	s.currentTable = s.currentTable[:len(s.currentTable)-1]
}

func (s *SqlVisitor) currentTableId() int {
	return s.currentTable[len(s.currentTable)-1]
}

func (s *SqlVisitor) addParam(p any) int {
	s.params = append(s.params, p)
	return len(s.params)
}

func (s *SqlVisitor) onTableFilter(f *TableFilter) {
	f.FilterGroup.Accept(s)
}

func (s *SqlVisitor) onFilterGroup(f *FilterGroup) {
	writeBrackets := (len(f.Filters) + len(f.Groups)) > 1
	if writeBrackets {
		s.query.WriteString("(")
	}

	for idx := range f.Filters {
		if idx > 0 {
			s.query.WriteString(" ")
			s.query.WriteString(string(f.Combinator))
			s.query.WriteString(" ")
		}

		f.Filters[idx].Accept(s)
	}

	for idx := range f.Groups {
		if idx > 0 || len(f.Filters) > 0 {
			s.query.WriteString(" ")
			s.query.WriteString(string(f.Combinator))
			s.query.WriteString(" ")
		}

		f.Groups[idx].Accept(s)
	}

	if writeBrackets {
		s.query.WriteString(")")
	}
}

func (s *SqlVisitor) onValueFilter(f *ValueFilter) {
	s.query.WriteString(fmt.Sprintf("T%d.%s ", s.currentTableId(), f.Column))

	switch f.Operation {
	case Eq:
		{
			param := s.addParam(f.Value)
			s.query.WriteString(fmt.Sprintf("= $%d", param))
		}
	case Neq:
		{
			param := s.addParam(f.Value)
			s.query.WriteString(fmt.Sprintf("<> $%d", param))
		}
	case Lt:
		{
			param := s.addParam(f.Value)
			s.query.WriteString(fmt.Sprintf("< $%d", param))
		}
	case Gt:
		{
			param := s.addParam(f.Value)
			s.query.WriteString(fmt.Sprintf("> $%d", param))
		}
	case Lte:
		{
			param := s.addParam(f.Value)
			s.query.WriteString(fmt.Sprintf("<= $%d", param))
		}
	case Gte:
		{
			param := s.addParam(f.Value)
			s.query.WriteString(fmt.Sprintf(">= $%d", param))
		}
	case In:
		{
			param := s.addParam(f.Value)
			s.query.WriteString(fmt.Sprintf("= ANY($%d)", param))
		}
	case NotIn:
		{
			param := s.addParam(f.Value)
			s.query.WriteString(fmt.Sprintf("<> ANY($%d)", param))
		}
	case IsNull:
		{
			s.query.WriteString("IS NULL")
		}
	case IsNotNull:
		{
			s.query.WriteString("IS NOT NULL")
		}
	case Contains:
		{
			param := s.addParam(f.Value)
			s.query.WriteString(fmt.Sprintf("LIKE '%%' || $%d || '%%'", param))
		}
	default:
		panic(fmt.Sprintf("operation %d is not supported", f.Operation))
	}
}

func (s *SqlVisitor) onJoinColumnFilter(f *JoinColumnFilter) {
	parentTableId := s.currentTableId()
	childTableId := s.pushNewTable()

	s.query.WriteString(fmt.Sprintf(
		"EXISTS(SELECT TRUE FROM %s AS T%d WHERE T%d.%s = T%d.%s",
		f.Filter.Table,
		childTableId,
		childTableId,
		f.ChildColumn,
		parentTableId,
		f.ParentColumn,
	))

	if !f.Filter.FilterGroup.IsEmpty() {
		s.query.WriteString(" AND ")
		f.Filter.Accept(s)
	}

	s.query.WriteString(")")
	s.popOldTable()
}

func (s *SqlVisitor) onJoinTableFilter(f *JoinTableFilter) {
	parentTableId := s.currentTableId()
	joinTableId := s.pushNewTable()
	childTableId := s.pushNewTable()

	s.query.WriteString(fmt.Sprintf(
		"EXISTS(SELECT TRUE FROM %s AS T%d INNER JOIN %s AS T%d ON T%d.%s = T%d.%s WHERE T%d.%s = T%d.%s",
		f.JoinTable,
		joinTableId,
		f.ChildTableFilter.Table,
		childTableId,
		childTableId,
		f.ChildTableJoinColumn,
		joinTableId,
		f.JoinTableChildColumn,
		joinTableId,
		f.JoinTableParentColumn,
		parentTableId,
		f.ParentTableJoinColumn,
	))

	if !f.ChildTableFilter.FilterGroup.IsEmpty() {
		s.query.WriteString(" AND ")
		f.ChildTableFilter.Accept(s)
	}

	s.query.WriteString(")")
	s.popOldTable()
	s.popOldTable()
}

File: graphql-rust/src/query/sql_builder.rs
use crate::query::filter_group::FilterGroup;
use crate::query::filter_type::FilterType;
use crate::query::join_column_filter::JoinColumnFilter;
use crate::query::join_table_filter::JoinTableFilter;
use crate::query::joined_table::JoinedTable;
use crate::query::ops::Operation;
use crate::query::table_filter::TableFilter;
use crate::query::value_filter::ValueFilter;
use crate::query::visitor::Visitor;
use std::fmt::Write;
use tokio_postgres::types::ToSql;

const SQL_QUERY_BUFFER_INITIAL_CAPACITY: usize = 1024;

#[derive(Debug, Default)]
pub struct SqlVisitor<'v> {
    joined_table: Option<JoinedTable<'v>>,
    next_table: u32,
    current_table: Vec<u32>,
    query: String,
    params: Vec<&'v (dyn ToSql + Sync)>,
}

impl<'v> SqlVisitor<'v> {
    pub fn new() -> Self {
        Default::default()
    }

    pub fn with_joined_table(joined_table: JoinedTable<'v>) -> Self {
        Self {
            joined_table: Some(joined_table),
            ..Default::default()
        }
    }

    pub fn translate(
        mut self,
        filter: &TableFilter<'v>,
        select: &[&str],
    ) -> (String, Vec<&'v (dyn ToSql + Sync)>) {
        self.query.reserve(SQL_QUERY_BUFFER_INITIAL_CAPACITY);

        let current_table_id = self.push_new_table();

        self.query.push_str("SELECT ");
        if select.is_empty() {
            write!(&mut self.query, "T{}.*", current_table_id).unwrap();
        } else {
            for (idx, sel) in select.iter().enumerate() {
                if idx > 0 {
                    self.query.push_str(", ");
                }
                write!(&mut self.query, "T{}.{}", current_table_id, sel).unwrap();
            }
        }

        self.query.push_str(" FROM ");
        if let Some(j) = self.joined_table.as_ref() {
            write!(
                &mut self.query,
                "{} AS J INNER JOIN {} AS T{} ON J.{} = T{}.{}",
                j.join_table,
                filter.table_name(),
                current_table_id,
                j.join_table_join_col,
                current_table_id,
                j.data_table_join_col
            )
            .unwrap();
        } else {
            write!(
                &mut self.query,
                "{} AS T{}",
                filter.table_name(),
                current_table_id
            )
            .unwrap();
        }

        if !filter.filter_group().is_empty() || self.joined_table.is_some() {
            self.query.push_str(" WHERE ");
        }

        if !filter.filter_group().is_empty() {
            filter.accept(&mut self);
        }

        if let Some(j) = self.joined_table.take() {
            if !filter.filter_group().is_empty() {
                self.query.push_str(" AND ");
            }

            let param = self.add_param(j.join_table_filter_col_val);
            write!(
                &mut self.query,
                "J.{} = ${}",
                j.join_table_filter_col, param
            )
            .unwrap();
        }

        (self.query, self.params)
    }

    fn add_param(&mut self, p: &'v (dyn ToSql + Sync)) -> usize {
        self.params.push(p);
        self.params.len()
    }

    fn push_new_table(&mut self) -> u32 {
        self.current_table.push(self.next_table);
        self.next_table += 1;
        self.next_table - 1
    }

    fn pop_old_table(&mut self) {
        self.current_table.pop();
    }

    fn current_table_id(&self) -> u32 {
        self.current_table.last().copied().unwrap()
    }
}

impl<'v> Visitor<'v> for SqlVisitor<'v> {
    fn on_table_filter(&mut self, f: &TableFilter<'v>) {
        f.filter_group().accept(self);
    }

    fn on_filter_group(&mut self, f: &FilterGroup<'v>) {
        let write_brackets = (f.filters().len() + f.groups().len()) > 1;
        if write_brackets {
            self.query.push_str("(");
        }

        for (idx, filter) in f.filters().iter().enumerate() {
            if idx > 0 {
                write!(&mut self.query, " {} ", f.combinator()).unwrap();
            }

            filter.accept(self);
        }

        for (idx, group) in f.groups().iter().enumerate() {
            if idx > 0 || !f.filters().is_empty() {
                write!(&mut self.query, " {} ", f.combinator()).unwrap();
            }

            group.accept(self);
        }

        if write_brackets {
            self.query.push_str(")");
        }
    }

    fn on_filter_type(&mut self, f: &FilterType<'v>) {
        match f {
            FilterType::ValueFilter(x) => x.accept(self),
            FilterType::JoinColumnFilter(x) => x.accept(self),
            FilterType::JoinTableFilter(x) => x.accept(self),
        }
    }

    fn on_value_filter(&mut self, f: &ValueFilter<'v>) {
        let current_table_id = self.current_table_id();

        write!(&mut self.query, "T{}.{} ", current_table_id, f.column()).unwrap();

        let val = f.value();

        match f.operation() {
            Operation::Eq => {
                let param = self.add_param(val);
                write!(&mut self.query, "= ${}", param)
            }
            Operation::Neq => {
                let param = self.add_param(val);
                write!(&mut self.query, "<> ${}", param)
            }
            Operation::Lt => {
                let param = self.add_param(val);
                write!(&mut self.query, "< ${}", param)
            }
            Operation::Gt => {
                let param = self.add_param(val);
                write!(&mut self.query, "> ${}", param)
            }
            Operation::Lte => {
                let param = self.add_param(val);
                write!(&mut self.query, "<= ${}", param)
            }
            Operation::Gte => {
                let param = self.add_param(val);
                write!(&mut self.query, ">= ${}", param)
            }
            Operation::In => {
                let param = self.add_param(val);
                write!(&mut self.query, "= ANY (${})", param)
            }
            Operation::NotIn => {
                let param = self.add_param(val);
                write!(&mut self.query, "<> ANY (${})", param)
            }
            Operation::IsNull => {
                write!(&mut self.query, "IS NULL")
            }
            Operation::IsNotNull => {
                write!(&mut self.query, "IS NOT NULL")
            }
            Operation::Contains => {
                let param = self.add_param(val);
                write!(&mut self.query, "LIKE '%' || ${} || '%'", param)
            }
        }
        .unwrap();
    }

    fn on_join_column_filter(&mut self, f: &JoinColumnFilter<'v>) {
        let parent_table_id = self.current_table_id();
        let child_table_id = self.push_new_table();

        write!(
            &mut self.query,
            "EXISTS(SELECT TRUE FROM {} AS T{} WHERE T{}.{} = T{}.{}",
            f.filter().table_name(),
            child_table_id,
            child_table_id,
            f.child_column(),
            parent_table_id,
            f.parent_column()
        )
        .unwrap();

        if !f.filter().filter_group().is_empty() {
            self.query.push_str(" AND ");
            f.filter().accept(self);
        }

        self.query.push_str(")");
        self.pop_old_table();
    }

    fn on_join_table_filter(&mut self, f: &JoinTableFilter<'v>) {
        let parent_table_id = self.current_table_id();
        let join_table_id = self.push_new_table();
        let child_table_id = self.push_new_table();

        write!(
            &mut self.query,
            "EXISTS(SELECT TRUE FROM {} AS T{} INNER JOIN {} AS T{} ON T{}.{} = T{}.{} WHERE T{}.{} = T{}.{}",
            f.join_table(),
            join_table_id,
            f.filter().table_name(),
            child_table_id,
            child_table_id,
            f.child_table_join_col(),
            join_table_id,
            f.join_table_child_col(),
            join_table_id,
            f.join_table_parent_col(),
            parent_table_id,
            f.parent_table_join_col()
        ).unwrap();

        if !f.filter().filter_group().is_empty() {
            self.query.push_str(" AND ");
            f.filter().accept(self);
        }

        self.query.push_str(")");
        self.pop_old_table();
        self.pop_old_table();
    }
}
